### ISSUE TITLE
Will a pen text block be added into Scratch?

### ISSUE THREAD

**Issue by cyx2015s** [LLK/scratch-gui#5714]
Will a pen text block be added into Scratch?

If it can be added, pen projects will be much 
more simpler.

(Screenshot taken from another website)
![IMG_20200507_062746](https://user-images.githubusercontent.com/31377746/81234753-f6fd2480-902b-11ea-86e8-779f48b7a33a.jpg)

**Comment by BoomerScratch** [comment]
No, i don't think so. You do have one in [Snap](http://snap.berkeley.edu):
![image](https://user-images.githubusercontent.com/60622217/81276458-53276f00-9053-11ea-8d78-055e67e3bdea.png)

**Comment by benjiwheeler** [comment]
There has been some thinking about this over the years, but there has never been a design proposal that everyone felt fit just right. Proposals for how this would work are welcome. How would positioning work? How would font choice work? How would size work? How would multiple lines work?

**Comment by griffpatch** [comment]
You could go down the route of adding it to pen, using pen width to set the
size of the font (relative). Or you could add a text extension that have
you more control including font size, alignment and text width reporters.

To be honest, of prefer the latter, and it might be worth considering
allowing the text later to be able to be at the front of all other
layers...? As you say it's a hard call though as there are so many use
cases that vary what works be most useful. If the text is limited to pen
then it will look bad...

How about a completely new idea, a scratch block that changes the costume
of the current sprite into the vector costume of a given text string for
example? Then we could create clones, order then, move them, rotate them,
add effects... Could be great! Just what I need lol :)

**Comment by griffpatch** [comment]
So... That costume idea, when you use the set text costume to "griffpatch",
the current sprite costume is generated in the fly too the text
"griffpatch". I can size and position and even stamp it if required. Then
if I use "set costume" afterwards the sprite will return to the normal
costumes (and the text costume will cease to exist). You would still need
to consider how to select font, and text alignment, etc

**Comment by apple502j** [comment]
Here's my idea:
Text Extension
`write []` - write the text using current settings. the text won't be updated when they change.
`set color to []` - set color using color
`set () to ()` - set width and color using numeric value like Pen
`change () by ()` - see above
`set font to ()` - setting the font, the option is just paint editor's
`width of []` - calculating width of the text. this may be slow but who knows
`clear all text` - clearing all text

position: the sprite's X and Y is used
alignment: I think, with all those blocks above, it's possible to do it user-side.
layer: I think the front layer is good, but maybe we can add options to change it to "between stage and pen", or "under pen".

**Comment by JeffaloBob** [comment]
this is something that i would really like to see, it seems like a very simple feature that would really lower the bar of entry for some really nice things. right now if you want to print custom text, you'll have to use someone's PTE or CTE but they're not exactly very efficient ot easy to use as something built into scratch.

people can be more creative if they can print text to the screen, right now it seems like a really odd limitation of scratch.

**Comment by benjiwheeler** [comment]
I think it would be useful to be able to revise the text's content without having to redraw any overlapping pen drawing. That would require that a piece of text be kept track of as an instance.

One way to do this would be:
1. introduce a positionable text extension
2. when positionable text extension is present, each sprite (and each clone) can have up to one text instance
3. the text would be located at the sprite's location, so all move blocks, color blocks and size blocks would work as normal
3. extension blocks would exist to change that sprite's text's content, font, and justification (left/center/right)
4. multi-line text could also be possible, if blocks are added to set line height and margin width

**Comment by ericrosenbaum** [comment]
I'm interested in this as well (and have thought about it for a long time). I'm hoping to do some prototyping so we can compare these different concepts. My guess is that the idea where a sprite's costume becomes text (or the text is layered onto it) will be the one that feels simplest and most powerful. Even though it is different from existing PTEs, it should be more accessible to beginners, and still afford most or all of the same flexibility. I think! We gotta try it...

**Comment by Explosion-Scratch** [comment]
I think what @griffpatch suggested sounds like a great idea! With this there would be very precise controll of things like text wrapping, size, color etc. in an intuitive way.

**Comment by Kenny2github** [comment]
@benjiwheeler Just noting that if multi-line text is supported, vertical align comes into play too (in addition to justification).

**Comment by cyx2015s** [comment]
I think implementing a text extension (either with pen or with costume) will make Scratch projects much more expandable, lighter and faster. Recently I saw a project called Ivis's Fantasy, in order to display the conversation char by char, he created more than 800 costumes(Of course he did it by using a program to add text costumes automatically). He is a Chinese Scratcher. Lastly, I think both pen text and costume text are important.

**Comment by griffpatch** [comment]
If you have costume text, then you should be able to stamp which gives you
the pen text for free... which is rather neat.
One note, if the text was to be 'added' to an existing costume, the
downside is that 'touching' blocks would then report they were touching the
sprite if they touched the text... but I guess there are issues with that
in all sprite based text ideas. It would be super nice if scratch had an
'exclude from touching' toggle feature on sprites so you could make clones
and opt in or out of touching per clone... but that's another issue :)

**Comment by Explosion-Scratch** [comment]
@griffpatch. The problem with adding toggles and setting such as alignment, exclude from touching, etc. is that eventually the system gets to complex and confusing for scratchers, especially new scratchers. Another problem with the costume based approach is that when using the stamp block the quality of the stamp is much lower than that of the sprite. Therefre in my opinion a costume based approach is still better as it keeps scratch being intuitive while still adding features that if added in a different manner could be seen as advanced and/or confusing. Also for the touching problem could you use broadcasts and a run with out screen refresh custom block to wuickly hide, sense then show again?

**Comment by GrahamSH-LLK** [comment]
@cyx2015s By the way, what website is that?

**Comment by cyx2015s** [comment]
> @cyx2015s By the way, what website is that?

www.aerfaying.com

**Comment by redstone2010** [comment]
Cool idea, I support 👍🏻

**Comment by Explosion-Scratch** [comment]
Same here! :) This sounds cool.
>
>

**Comment by hyn-2020** [comment]
Look forward to this new feature.

**Comment by griffpatch** [comment]
Woo-hoo!!! https://lab.scratch.mit.edu/text/ Loving seeing the text costume idea implemented in Scratch Labs - I think it's brilliant and exactly what I was getting at :)

Here's my feedback:
* Its awesome - we need it :)
* Should the rotation centre be around the center of the multiline costume rather than the center of the top line? Kids will rotate their text to zoom it in and it will look much better if rotated around the real centre... unless you can also set the alignment of the vertical as well as the horizontal. I'm torn here.
* Can we have a reported that gives back the width and height of the text? Super useful when you want to work with text in a project (of course you can use the fencing trick to calculate sizes I guess but worth considering.
* One more idea that springs to mind - Have you considered adding an effect like 'shadow' or 'outline' as sprite effects because that would really complement both this extension and existing sprites. Using the text sprites in clones will be super useful if we can use one clone per bit of text, but to make text stand out in a busy screen it is very rarely just plain text but has a contrasty outline (light or dark) or a shadow / glow. For this you need to be able to draw an outlined version of the text somehow.

Thanks for looking at adding this amazing feature!

Griffpatch

**Comment by ericrosenbaum** [comment]
Thanks for your feedback @griffpatch! 

re: rotation center of multiline text, I see your point about rotation. I was thinking about the use case of a "text box", where you have a fixed region where you want to display the text. In that case I think it's better to put the rotation center on the top line, so you can consistently position the text no matter how many lines you have.

re: size reporters, can you give an example of a use case there?

re: shadow and outline effects, I agree it would be great to have something like this. I experimented with adding outlines, but some of our fonts do not render them correctly (you can see the same bug when adding outlines to text in the paint editor- you get weird spiky artifacts). I also worked a bit on adding something more like a "highlight", a colored box behind the text, which is another way to address the problem of readability on busy backgrounds- I just ran out of time on that.

**Comment by griffpatch** [comment]
Rotation Centers: I do agree that the positioning of the text as it is does
make sense for a lot of scenarios - Which is why I wondered whether you
might need a way of changing the vertical alignment just as you can the
horizontal alignment?

Size Reporters: The case for this is where you want to have one word in a
different color, or include icons in your text, etc so you can write some
text, clone then move the sprite by the width of the text and change the
costume, etc... This might have it's place though as a global feature that
could be useful - but TBH would not be a massive loss not to have, just
more so in this case as we've never had a costume that we didn't otherwise
know the width of before like this - and possibly more importantly the
height as it can wrap and cause us potential layout issues...

Shadow/Outline: Yeah, I've seen the problems with giving our fonts outlines
(extended pointy corners). Is it possible we need to apply a different
stroke line joining mode?
https://developer.mozilla.org/en-US/docs/Web/SVG/Attribute/stroke-linejoin
- It is also worth adding that when I use outlines like this I would
generally need the outline to be behind the fill so that it doesn't eat
into the font text itself (fat outline half outside the font, half inside
so the inner letter shape is less discernible). I think this can be done in
SVG. Re using a background highlight, I'm not so keen, although it is one
solution :)

I do hope more time can be scheduled for this amazing new feature - It will
make a lot of Scratchers VERY happy - The community has been buzzing about
it... There are a lot of calls for integration with google fonts ;) lol

**Comment by cyx2015s** [comment]
404

**Comment by Secret-chest** [comment]
I suggest removing the animations, they inhibit creativity.

**Comment by benjiwheeler** [comment]
One thing I keep wishing for, when playing with this: the ability to show text AND the sprite's costume, at the same time. Then you could, for example, easily create a button as a single sprite, and have it change size and color and other effects always in sync.

**Comment by griffpatch** [comment]
True, but then having to use the set sprite size to change font size would
need to be independent?

**Comment by Secret-chest** [comment]
> I suggest removing the animations, they inhibit creativity.

Because you can easily make them yourself.

**Comment by mxmou** [comment]
> > I suggest removing the animations, they inhibit creativity.
> 
> Because you can easily make them yourself.

I don't think there's an easy way to implement the rainbow animation without the block. The workarounds for type and zoom are quite simple but they might not be obvious to new users.

**Comment by Secret-chest** [comment]
> > > I suggest removing the animations, they inhibit creativity.
> > 
> > Because you can easily make them yourself.
> 
> I don't think there's an easy way to implement the rainbow animation without the block. The workarounds for type and zoom are quite simple but they might not be obvious to new users.

I still think such a specific block is not needed.

**Comment by Secret-chest** [comment]
> > > I suggest removing the animations, they inhibit creativity.
> > 
> > Because you can easily make them yourself.
> 
> I don't think there's an easy way to implement the rainbow animation without the block. The workarounds for type and zoom are quite simple but they might not be obvious to new users.

The same users should learn about loops.
